common/djangoapps/third_party_auth/tests/specs/test_testshib.py

```diff
@@ -155,41 +203,154 @@ def test_configure_testshib_provider_with_cache_duration(self):
+
+        httpretty.register_uri(httpretty.POST, SAPSF_ASSERTION_URL, content_type='text/plain', body=assertion_callback)
+
+        SAPSF_BAD_ASSERTION_URL = 'http://successfactors.com/oauth-fake/idp'
+
+        def bad_callback(_request, _uri, headers):
+            """
+            Return a 404 error when someone tries to call the URL.
+            """
+            return (404, headers, 'NOT AN ASSERTION')
+
+        httpretty.register_uri(httpretty.POST, SAPSF_BAD_ASSERTION_URL, content_type='text/plain', body=bad_callback)
+
+        # Mock the call to the SAP SuccessFactors token endpoint
+        SAPSF_TOKEN_URL = 'http://successfactors.com/oauth/token'
+
+        def token_callback(_request, _uri, headers):
+            """
+            Return a fake assertion after checking that the input is what we expect.
+            """
+            self.assertIn('assertion=fake_saml_assertion', _request.body)
+            self.assertIn('company_id=NCC1701D', _request.body)
+            self.assertIn('grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Asaml2-bearer', _request.body)
+            self.assertIn('client_id=TatVotSEiCMteSNWtSOnLanCtBGwNhGB', _request.body)
+            return (200, headers, '{"access_token": "faketoken"}')
+
+        httpretty.register_uri(
+            httpretty.POST,
+            SAPSF_TOKEN_URL,
+            content_type='application/json',
+            body=token_callback
         )
+
+        # Mock the call to the SAP SuccessFactors OData user endpoint
+        ODATA_USER_URL = (
+            'http://api.successfactors.com/odata/v2/User(userId=\'myself\')'
+            '?$select=username,firstName,lastName,defaultFullName,email'
+        )
+
+        def user_callback(_request, _uri, headers):
+            auth_header = headers.get('Authorization')
+            self.assertEqual(auth_header, 'Bearer faketoken')
+            return (
+                200,
+                headers,
+                json.dumps({
+                    'd': {
+                        'username': 'jsmith',
+                        'firstName': 'John',
+                        'lastname': 'Smith',
+                        'defaultFullName': 'John Smith',
+                        'email': 'john@smith.com',
+                    }
+                })
+            )
+
+        httpretty.register_uri(httpretty.GET, ODATA_USER_URL, content_type='application/json', body=user_callback)
+
+    def test_register_insufficient_sapsf_metadata(self):
+        """
+        Configure the provider such that it doesn't have enough details to contact the SAP
+        SuccessFactors API, and falls back to the data it receives from the SAML assertion.
+        """
+        self._configure_testshib_provider(
+            identity_provider_type='sap_success_factors',
+            other_settings='{"key_i_dont_need":"value_i_also_dont_need"}',
+        )
+        # Because we're getting details from the assertion, fall back to the initial set of details.
```

Comment:
@haikuginger It would be worth duplicating this comment below, I think (above L341).